src/Server.php

```diff
@@ -226,7 +238,27 @@ public function handleRequest(ConnectionInterface $conn, RequestInterface $reque
         );
 
         $callback = $this->callback;
-        $callback($request, $response);
+        $promise = new Promise(function ($resolve, $reject) use ($callback, $request) {
+            $resolve($callback($request));
+        });
+
+        $that = $this;
+        $promise->then(
+            function ($response) use ($that, $conn, $request) {
+                if (!$response instanceof ResponseInterface) {
+                    $that->emit('error', array(new \InvalidArgumentException('Invalid response type')));
+                    return $that->writeError($conn, 500);
+                }
+                $that->handleResponse($conn, $response, $request->getProtocolVersion());
+            },
+            function ($ex) use ($that, $conn) {
+                if (!$ex instanceof \Exception) {
+                    $ex = new \InvalidArgumentException('Unknown rejection type');
```

Comment:
Very generic exception message. Suggestion:

```php
$ex = new \InvalidArgumentException(
     sprintf(
        'The response callback returned a rejected promise with a reason of type "%s" instead of a \Throwable or \Exception instance.'
        is_object($ex) ? get_class($ex) : gettype($ex)
    )
);
```